xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,345 @@
+      return;
+    }
+    AtomicReference<ConnectivityStateInfo> subchannelStateRef =
+        getSubchannelStateInfoRef(subchannel);
+    if (!subchannelStateRef.get().getState().equals(READY) && stateInfo.getState().equals(READY)) {
+      subchannelEnteredReady = true;
+    }
+
+    // Do not proactively reconnect even if the subchannel is previously connected.
+    subchannelStateRef.set(stateInfo);
+    updateBalancingState(currentPicker);
+  }
+
+  private static AtomicReference<ConnectivityStateInfo> getSubchannelStateInfoRef(
+      Subchannel subchannel) {
+    return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO");
+  }
+
+  private final class RingHashPicker extends SubchannelPicker {
+    private final List<RingEntry> ring;
+
+    private RingHashPicker(List<RingEntry> ring) {
+      this.ring = ring;
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      long requestHash = args.getCallOptions().getOption(XdsNameResolver.RPC_HASH_KEY);
+      if (ring.isEmpty()) {
+        return PickResult.withNoResult();
+      }
+      // Find the nearest corresponding host clockwise around the ring.
+      int low = 0;
+      int high = ring.size();
+      int mid;
+      while (true) {
+        mid = (low + high) / 2;
+        if (mid == ring.size()) {
+          mid = 0;
+          break;
+        }
+        long midVal = ring.get(mid).hash;
+        long midValL = mid == 0 ? 0 : ring.get(mid - 1).hash;
+        if (requestHash <= midVal && requestHash > midValL) {
+          break;
+        }
+        if (midVal < requestHash) {
+          low = mid + 1;
+        } else {
+          high =  mid - 1;
+        }
+        if (low > high) {
+          mid = 0;
+          break;
+        }
+      }
+      EquivalentAddressGroup serverAddr = ring.get(mid).eag;
+      EquivalentAddressGroup addrKey = stripAttrs(serverAddr);
+      Subchannel subchannel = subchannels.get(addrKey);
+      if (subchannel != null && getSubchannelStateInfoRef(subchannel).get().getState() == READY) {
+        return PickResult.withSubchannel(subchannel);
+      }
+      if (subchannel == null) {
+        Attributes attr = Attributes.newBuilder().set(
+            STATE_INFO, new AtomicReference<>(ConnectivityStateInfo.forNonError(IDLE))).build();
+        subchannel = helper.createSubchannel(
+            CreateSubchannelArgs.newBuilder()
+                .setAddresses(serverAddr).setAttributes(attr).build());
+        final Subchannel finalSubchannel = subchannel;
+        subchannel.start(new SubchannelStateListener() {
+          @Override
+          public void onSubchannelState(ConnectivityStateInfo newState) {
+            processSubchannelState(finalSubchannel, newState);
+          }
+        });
+        subchannels.put(addrKey, subchannel);
+      }
+      subchannel.requestConnection();
+      // Return no result to queue the pick and re-attempt later.
+      return PickResult.withNoResult();
```

Comment:
FIXME: there is a bug here, which may cause the RPC to hang forever if the selected subchannel can never successfully connct.

Reply:
Fixed LB-level subchannel state management (stay in TF until being READY) and fall back to try other subchannels (clockwise in the ring) if the one hit by hash is in TF.